flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorassembler/VectorAssembler.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A feature transformer that combines a given list of input columns into a vector column. Types of
+ * input columns must be either vector or numerical value.
+ */
+public class VectorAssembler
+        implements Transformer<VectorAssembler>, VectorAssemblerParams<VectorAssembler> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+    private static final double RATIO = 1.5;
+
+    public VectorAssembler() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(
+                                inputTypeInfo.getFieldTypes(), TypeInformation.of(Vector.class)),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
+        DataStream<Row> output =
+                tEnv.toDataStream(inputs[0])
+                        .flatMap(
+                                new AssemblerFunc(getInputCols(), getHandleInvalid()),
+                                outputTypeInfo);
+        Table outputTable = tEnv.fromDataStream(output);
+        return new Table[] {outputTable};
+    }
+
+    private static class AssemblerFunc implements FlatMapFunction<Row, Row> {
+        private final String[] inputCols;
+        private final String handleInvalid;
+
+        public AssemblerFunc(String[] inputCols, String handleInvalid) {
+            this.inputCols = inputCols;
+            this.handleInvalid = handleInvalid;
+        }
+
+        @Override
+        public void flatMap(Row value, Collector<Row> out) throws Exception {
+            Object[] objects = new Object[inputCols.length];
+            for (int i = 0; i < objects.length; ++i) {
+                objects[i] = value.getField(inputCols[i]);
+            }
+            Vector assembledVector = null;
+            try {
+                assembledVector = assemble(objects);
+            } catch (Exception e) {
+                switch (handleInvalid) {
+                    case HasHandleInvalid.ERROR_INVALID:
+                        throw e;
+                    case HasHandleInvalid.SKIP_INVALID:
+                        return;
+                    case HasHandleInvalid.KEEP_INVALID:
```

Comment:
When `keep` is specified for this transformer, do we keep all types of invalidity (e.g. missing column, types that are neither number nor vector), or just a selected list of invalid types?

And what is the expected output of the transform(...) if it contains invalid input rows? For example, the output row's outputColumn is null?

If we plan to keep only a selected types of `invalidity`, as it affects the algorithm behavior, we probably need to explicitly specify it in the Java doc of this class.

Reply:
Here，keep is just keep this input row values，the outputColumn has no meaning，so I give a null.

I have refine code for keep condition, including missing column, types err. and selected invalid types.

I add some doc for keep invalid.

Reply:
Thanks for the update. Since the method here is an implementation detail of `VectorAssembler`, it is probably better to move the Java doc to the `HANDLE_INVALID` parameter in the `VectorAssemblerParams`. And we probably also need to explain the semantics of `error` and `skip`.

How about using the following Java doc for the `HANDLE_INVALID` parameter?

```
    /**
     * <p>Supported options and the corresponding behavior to handle invalid entries is listed as
     * follows.
     *
     * <ul>
     *   <li>error: raise an exception.
     *   <li>skip: filter out rows with bad values.
     *   <li>keep: output bad rows with output column's value set to null.
     * </ul>
     */
```

Reply:
I agree with it.